pep-0544.txt

```diff
@@ -439,98 +448,123 @@ but a static type checker will handle them specially. Subclassing a protocol
 
   from typing import Sized, Protocol
 
-  class SizedAndCloseable(Sized, Protocol):
+  class SizedAndClosable(Sized, Protocol):
       def close(self) -> None:
           ...
 
-Now the protocol ``SizedAndCloseable`` is a protocol with two methods,
+Now the protocol ``SizedAndClosable`` is a protocol with two methods,
 ``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
 this would be a regular (non-protocol) class that must implement ``Sized``.
-If ``Protocol`` is included in the base class list, all the other base classes
-must be protocols. A protocol can't extend a regular class.
-
-Alternatively, one can implement ``SizedAndCloseable`` like this, assuming
-the existence of ``SupportsClose`` from the example in `definition`_ section::
+Alternatively, one can implement ``SizedAndClosable`` protocol by merging
+the ``SupportsClose`` protocol from the example in `definition`_ section
+with ``typing.Sized``::
 
   from typing import Sized
 
-  class SupportsClose(...): ...  # Like above
+  class SupportsClose(Protocol):
+      def close(self) -> None:
+          ...
 
-  class SizedAndCloseable(Sized, SupportsClose, Protocol):
+  class SizedAndClosable(Sized, SupportsClose, Protocol):
       pass
 
 The two definitions of ``SizedAndClosable`` are equivalent.
 Subclass relationships between protocols are not meaningful when
 considering subtyping, since structural compatibility is
 the criterion, not the MRO.
 
-Note that rules around explicit subclassing are different from regular ABCs,
-where abstractness is simply defined by having at least one abstract method
-being unimplemented. Protocol classes must be marked *explicitly*.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class, see `rejected`_
+ideas for rationale. Note that rules around explicit subclassing are different
+from regular ABCs, where abstractness is simply defined by having at least one
+abstract method being unimplemented. Protocol classes must be marked
+*explicitly*.
 
 
-Generic and recursive protocols
--------------------------------
+Generic protocols
+-----------------
 
 Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
 and ``Iterator`` are generic protocols. They are defined similar to normal
 non-protocol generic types::
 
-  T = TypeVar('T', covariant=True)
-
   class Iterable(Protocol[T]):
       @abstractmethod
       def __iter__(self) -> Iterator[T]:
           ...
 
-Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
-``Protocol, Generic[T, S, ...]``.
+``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``. Declaring variance is not necessary for
+protocol classes, since it can be inferred from a protocol definition.
+Example::
+
+  class Box(Protocol[T]):
+      content: T
+
+  x: Box[float]
+  y: Box[int]
+  x = y  # This is OK due to inferred covariance of the protocol 'Box'.
```

Comment:
@JukkaL 
Yes, ``Box`` should be invariant, completely forgot about "mutability" in my implementation. Fortunately, this is easy to fix, I will just check subclassing both ways for members with ``IS_SETTABLE`` flag (basically treating a settable member as a pair of getter/setter methods).
> And for consistency with normal classes, I'd prefer variance to be explicit.

OK, I agree, I will just remove the text mentioning that variance could be omitted.
> And what if I'd like to explicitly make something invariant for some reason, even if inference thinks it can be covariant?

Good question! I am not sure what to do here. Currently, I do two-step checking:
* First check nominal subtyping using user-declared variance, if it succeeds, then OK, otherwise go to next step.
* Try structural subtyping. Here I check compatibility of every member (types and flags like ``IS_SETTABLE``, ``IS_CLASSVAR``, etc) I think this is quite robust, so that the inferred variance should be reliable (unless I am missing something).

In principle, I probably could always use the user-declared variance, but this will have two downsides:
* 90% of users use invariant variables, so that protocols will be interpret as invariant. For nominal classes, this makes perfect sense, but for protocols, where we anyway do a lot of work checking all members, we could reliably infer variance, and making them all invariant will be quite sad.
* Less important, but this will also make implementation more complex. Currently, when I check ``is_subtype(C, Proto[int])``, I just substitute ``T`` with ``int`` in ``Proto`` and check all members. With user declared variance I will first need something like ``map_instance_to_supertype`` and then use variance.